sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala

```diff
@@ -388,6 +391,27 @@ class ParquetSchemaSuite extends ParquetSchemaTest {
     }
   }
 
+  test("SPARK-37191: Schema merging for DecimalType with different precision but the same scale") {
+    import testImplicits._
+
+    withTempPath { dir =>
+      val path = dir.getCanonicalPath
+
+      val data1 = spark.sparkContext.parallelize(Seq(Row(new BigDecimal("123456789.11"))), 1)
+      val schema1 = StructType(StructField("col", DecimalType(12, 2)) :: Nil)
+
+      val data2 = spark.sparkContext.parallelize(Seq(Row(new BigDecimal("1234567890000.11"))), 1)
+      val schema2 = StructType(StructField("col", DecimalType(17, 2)) :: Nil)
+
+      spark.createDataFrame(data1, schema1).write.parquet(path)
+      spark.createDataFrame(data2, schema2).write.mode("append").parquet(path)
+
+      val res = spark.read.option("mergeSchema", "true").parquet(path)
+      assert(res.schema("col").dataType == DecimalType(17, 2))
+      res.foreach(_ => ()) // must not throw exception
```

Comment:
shall we use `checkAnswer` to check the result?

Reply:
maybe we should put the test in `ParquetQuerySuite`, so that we can test all the parquet readers.

Reply:
ParquetSchemaSuite did not have `checkAnswer` so I left it as is originally. This has been updated since, I moved the test to ParquetQuerySuite and added `withAllParquetReaders` to test against all of the parquet readers.